### PR TITLE
Add scroll to top button

### PR DESCRIPTION
**Pull request by evulins**
Create scroll to the button which will be added to the show problem page

### CODE REVIEW

File: src/components/pages/ProblemsPage/index.tsx [hunk 1]
```diff
@@ -63,6 +64,7 @@ const ProblemsPage: FC = () => {
           ))
         )}
       </Wrapper>
+      <ScrollToTopButton />
```
Comment (franiakat) [inline src/components/pages/ProblemsPage/index.tsx:67]: In the end this won't be added on this page, but on ProblemDetailedPage, right? :)

File: src/components/atoms/ScrollToTopButton/styles.tsx [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+import { minDevice } from "../../../assets/styles/breakpoints";
+
+const ScrollButton = styled.button`
+  position: fixed;
+  left: 85%;
+  top: 10%;
+  width: fit-content;
+  background: none;
+  border: none;
+  overflow-y: scroll;
```
Comment (franiakat) [inline src/components/atoms/ScrollToTopButton/styles.tsx]: Is that needed? I think not necessarily, but am I missing something? :)

File: src/components/atoms/ScrollToTopButton/styles.tsx [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+`;
+
+const ArrowUpIcon = styled(ArrowUp)<ArrowUpIconStyleProps>`
+  display: ${(props) => props.display};
```
Comment (franiakat) [inline src/components/atoms/ScrollToTopButton/styles.tsx]: Is this needed to be configurable from the outside? 🤔

File: src/components/atoms/ScrollToTopButton/index.tsx [hunk 4]
```diff
@@ -0,0 +1,37 @@
+import { FC, useState } from "react";
+
+import * as S from "./styles";
+
+const ScrollToTopButton: FC = () => {
+  const [visible, setVisible] = useState(false);
+
+  const toggleVisible = () => {
```
Comment (franiakat) [inline src/components/atoms/ScrollToTopButton/index.tsx:9]: Right now this function is being called on every `scroll` event which is a lot, going up and down randomly on our problem index now gives me hundreds of them 🙃 Therefore we can throttle the execution of the function, in a simpler words, sort of delay it on given conditions: https://www.sitepoint.com/throttle-scroll-events/
You can take a look at this article here. ;)

File: src/components/atoms/ScrollToTopButton/index.test.tsx [hunk 5]
```diff
@@ -0,0 +1,28 @@
+import { fireEvent, screen } from "@testing-library/react";
+import userEvent from "@testing-library/user-event";
+
+import { testRenderer } from "../../../setupTests";
+import ScrollToTopButton from "./index";
+
+describe("<ScrollToTopButtom />", () => {
+  it("doesn't display ScrollButton when the page is render", async () => {
```
Comment (franiakat) [inline src/components/atoms/ScrollToTopButton/index.test.tsx]: :)
```suggestion
  it("doesn't display ScrollButton when the page is first rendered", async () => {
```

File: src/components/atoms/ScrollToTopButton/index.test.tsx [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    testRenderer(<ScrollToTopButton />);
+
+    expect(
+      screen.queryByTitle(/Ikona przewijania strony do początku/i)
+    ).not.toBeVisible();
+  });
+
+  // I think we need to test this case in the component when the button will be implemented
```
Comment (franiakat) [inline src/components/atoms/ScrollToTopButton/index.test.tsx:16]: Do you mean used in ProblemDetailedPage? :)

File: src/components/atoms/ScrollToTopButton/index.test.tsx [hunk 7]
```diff
@@ -0,0 +1,28 @@
+import { fireEvent, screen } from "@testing-library/react";
+import userEvent from "@testing-library/user-event";
+
+import { testRenderer } from "../../../setupTests";
+import ScrollToTopButton from "./index";
+
+describe("<ScrollToTopButtom />", () => {
```
Comment (franiakat) [inline src/components/atoms/ScrollToTopButton/index.test.tsx:7]: BTW around this topic of testing our new component, jest has its limitations (what we already know from other tests that can't be written here): https://github.com/testing-library/react-testing-library/issues/671
We may think of adding puppeteer or Cypress as KCD suggests 🙃

**Review by franiakat — CHANGES_REQUESTED** [review]
Looking good! 💃 I had a few questions/thoughts. :)

**Comment by franiakat** [comment]
Two additional thoughts:
- you can take a look at this article saying how to configure the css behaviour for people with `prefers-reduced-motion'` configuration
https://ashleemboyer.com/accessible-smooth-scroll-to-top-buttons-with-little-code
- the nasty DS_Store files got added 🙊 you can remove them.

File: src/components/pages/ProblemsPage/index.tsx (hunk 1 above)
Comment (evulins) [reply src/components/pages/ProblemsPage/index.tsx:67]: Of course, I put it here just to see and test the component :)

File: src/components/atoms/ScrollToTopButton/index.test.tsx (hunk 6 above)
Comment (evulins) [reply src/components/atoms/ScrollToTopButton/index.test.tsx:16]: Yes :)

File: src/components/atoms/ScrollToTopButton/index.test.tsx (hunk 7 above)
Comment (evulins) [reply src/components/atoms/ScrollToTopButton/index.test.tsx:7]: I saw this discussion earlier today (when i was googling how to test scrolling) and i want to talk about KCD suggestions

File: src/components/atoms/ScrollToTopButton/styles.tsx (hunk 3 above)
Comment (evulins) [reply src/components/atoms/ScrollToTopButton/styles.tsx]: This is this place when typescript was screaming at me :( We can discuss this during our daily :)

File: src/components/atoms/ScrollToTopButton/index.tsx (hunk 4 above)
Comment (evulins) [reply src/components/atoms/ScrollToTopButton/index.tsx:9]: Good point. After I did some reading  and testing some functions I chose to add loadash.throttle to our app  :)

**Comment by evulins** [comment]
> Two additional thoughts:
> 
> * you can take a look at this article saying how to configure the css behaviour for people with `prefers-reduced-motion'` configuration
>   https://ashleemboyer.com/accessible-smooth-scroll-to-top-buttons-with-little-code

We can manage focus when the button will be used in the ProblemDetailedPage

File: src/components/atoms/ScrollToTopButton/index.test.tsx (hunk 6 above)
Comment (bazelajoanna) [reply src/components/atoms/ScrollToTopButton/index.test.tsx:16]: I think we can test it here, but we need to render `Navbar`, `ProblemsPage` and `Footer`:
`testRenderer(
      <>
        <Navbar />
        <ProblemsPage />
        <Footer />
      </>
    )`

File: src/components/atoms/ScrollToTopButton/index.tsx [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+      <S.ArrowUpIcon
+        title="Ikona przewijania strony do początku"
+        onClick={scrollToTop}
+        display={visible ? "inline" : "none"}
+      />
+    </S.ScrollButton>
+  );
+};
+
+export default ScrollToTopButton;
```
Comment (bazelajoanna) [inline src/components/atoms/ScrollToTopButton/index.tsx:37]: 👏 👏 👏

File: src/components/atoms/ScrollToTopButton/index.test.tsx (hunk 6 above)
Comment (evulins) [reply src/components/atoms/ScrollToTopButton/index.test.tsx:16]: Exactly ;) IMO this is the same case as it was with input. We don't want to render other components than the exact one we  would like test.

**Review by bazelajoanna — COMMENTED** [review]
Very nice! 🎊

File: src/components/atoms/ScrollToTopButton/index.test.tsx (hunk 6 above)
Comment (franiakat) [reply src/components/atoms/ScrollToTopButton/index.test.tsx:16]: Yes, let's wait until then and we will move it to ProblemDetailedPage test. 👌

**Review by franiakat — APPROVED** [review]
Very good piece of code 🎖️
sqash

**Pull request merged**